### PR TITLE
Add Hook to store arbitrary python objects that are copied over in tls

### PR DESCRIPTION
**Pull request by eellison**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #92851
* __->__ #89169

For the cudagraphs implementation, we would like to reuse objects that are defined in python across the forward and backward. The backward is run in a different thread, so to handle this we add an api for copying over arbitrary python objects in pytorch's thread local state, in the same way that C++ objects are copied over currently.

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

File: torch/csrc/Module.cpp [hunk 1]
```diff
@@ -1500,6 +1501,23 @@ Call this whenever a new thread is created in order to propagate values from
...
+  });
+
+  py_module.def("_get_obj_in_tls", [](std::string key) -> py::handle {
+    auto safe_pyobject =
+        at::impl::ThreadLocalPythonObjects::get_state().get(key);
+    auto obj = safe_pyobject->ptr(getPyInterpreter());
+    return py::handle(obj);
+  });
+
+  py_module.def("_key_in_tls", [](std::string key) -> bool {
```
Comment (albanD) [inline torch/csrc/Module.cpp:1517]: nit: `_is_key_in_tls`

File: aten/src/ATen/ThreadLocalPythonObjects.h [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+#include <unordered_map>
+
+namespace at {
+namespace impl {
+
+struct TORCH_API ThreadLocalPythonObjects {
+  static void set(std::string key, std::shared_ptr<SafePyObject> value);
+  static const std::shared_ptr<SafePyObject>& get(std::string key);
+  static bool contains(std::string key);
+  static int64_t size();
```
Comment (albanD) [inline aten/src/ATen/ThreadLocalPythonObjects.h:14]: This is not actually used, you can remove it?

**Review by albanD — APPROVED** [review]
Implementation sounds good even though I don't really know if this is the right solution to your problem as it is not described in the PR description.

**Comment by ezyang** [comment]
We are still doing this right?

**Pull request closed without merging**